packages/theme-language-server-common/src/rename/providers/LiquidVariableRenameProvider.ts

```diff
@@ -0,0 +1,93 @@
+import { BaseRenameProvider } from '../BaseRenameProvider';
+import { DocumentManager } from '../../documents';
+import {
+  LiquidHtmlNode,
+  LiquidVariableLookup,
+  NodeTypes,
+  LiquidTag,
+  NamedTags,
+} from '@shopify/liquid-html-parser';
+import { Range } from 'vscode-languageserver';
+import {
+  PrepareRenameParams,
+  PrepareRenameResult,
+  RenameParams,
+  TextDocumentEdit,
+  TextEdit,
+  WorkspaceEdit,
+} from 'vscode-languageserver-protocol';
+import { visit } from '../../visitor';
+
+export class LiquidVariableRenameProvider implements BaseRenameProvider {
+  constructor(private documentManager: DocumentManager) {}
+
+  async prepare(
+    node: LiquidHtmlNode,
+    ancestors: LiquidHtmlNode[],
+    params: PrepareRenameParams,
+  ): Promise<null | PrepareRenameResult> {
+    const document = this.documentManager.get(params.textDocument.uri);
+    const textDocument = document?.textDocument;
+
+    if (!textDocument || !node || !ancestors) return null;
+    if (node.type !== NodeTypes.AssignMarkup && node.type !== NodeTypes.VariableLookup) return null;
+
+    return {
+      range: Range.create(
+        textDocument.positionAt(node.position.start),
+        textDocument.positionAt(node.position.end),
+      ),
+      placeholder: node.name || '',
+    };
+  }
+
+  async rename(
+    node: LiquidHtmlNode,
+    ancestors: LiquidHtmlNode[],
+    params: RenameParams,
+  ): Promise<null | WorkspaceEdit> {
+    const document = this.documentManager.get(params.textDocument.uri);
+    const textDocument = document?.textDocument;
+
+    if (!textDocument || !node || !ancestors) return null;
+    if (node.type !== NodeTypes.AssignMarkup && node.type !== NodeTypes.VariableLookup) return null;
+    if (document.ast instanceof Error) return null;
+
+    const oldName = node.name || '';
+    const ranges: Range[] = visit(document.ast, {
+      VariableLookup(node: LiquidVariableLookup): Range | undefined {
+        if (node.name === oldName) {
+          const range = Range.create(
+            textDocument.positionAt(node.position.start),
+            textDocument.positionAt(node.position.end),
+          );
+
+          return range;
+        }
+      },
+      LiquidTag(node: LiquidTag) {
+        if (node.name != NamedTags.assign) return;
```

Comment:
like increment and so on

Reply:
Instead of handling these case-by-case, what do you think about doing a whole-word match/replace inside any liquid blocks?

This seems safer than finding every edge case manually. Especially if we ever introduce new `NamedTags`.

Reply:
Ok did a few checks and we should be good. I haven't updated the tests yet, but the current solution should be good.


VariableLookup should handle:
- increment
- decrement
- form
- paginate
- capture
- layout
- for (variable after the `in` part)
- tablerow (variable after the `in` part)

ForMarkup should handle
- for (variable after the `for` part)
- tablerow (variable after the `tablerow` part)

The new change should also handle variable scopes. So if you have two different for-loops, and they both use `for x in products`, changing `x` in one for-loop shouldn't rename the other.